core/src/main/scala/org/bykn/bosatsu/TotalityCheck.scala

```diff
@@ -0,0 +1,589 @@
+         * tail has no splices, and is thus finite
+         *
+         * This is using the rule:
+         * [*_, rest] = [rest] | [_, *_, rest]
+         */
+        val zero = tail
+        val oneOrMore = Right(WildCard) :: lp
+        // If the left and right are disjoint,
+        // this creates a different representation
+        // of the left
+        (difference0List(zero, rp), difference0List(oneOrMore, rp))
+          .mapN {
+            case (zz, oo)
+              if leftIsSuperSet(zz, ListPat(zero)) &&
+                leftIsSuperSet(oo, ListPat(oneOrMore)) =>
+              ListPat(lp) :: Nil
+            case (zz, oo) => zz ::: oo
+          }
+      case (_, Left(_) :: rtail) if matchesEmpty(rtail) =>
+        // this is a total match
+        Right(Nil)
+      case (_, Left(_) :: _) =>
+        // we know the right can't end in Left since
+        // it starts with Left and the tail is not empty
+        // we can make progress:
+
+        difference0List(lp.reverse, rp.reverse)
+          .map(_.map {
+            case ListPat(diff) => ListPat(diff.reverse)
+          })
+    }
+
+  def difference0(left: Pattern[Cons, Type], right: Pattern[Cons, Type]): Res[Patterns] =
+    (left, right) match {
+      case (_, WildCard | Var(_)) => Right(Nil)
+      case (WildCard | Var(_), Literal(_)) =>
+        // the left is infinite, and the right is just one value
+        Right(left :: Nil)
+      case (WildCard | Var(_), _) if isTotal(right) == Right(true) =>
+        Right(Nil)
+      case (WildCard, listPat@ListPat(rp)) =>
+        // _ is the same as [*_] for well typed expressions
+        checkListPats(listPat :: Nil) *>
+          difference0List(Left(None) :: Nil, rp)
+      case (Var(v), listPat@ListPat(rp)) =>
+        // v is the same as [*v] for well typed expressions
+        checkListPats(listPat :: Nil) *>
+          difference0List(Left(Some(v)) :: Nil, rp)
+      case (left@ListPat(lp), right@ListPat(rp)) =>
+        checkListPats(left :: right :: Nil) *>
+          difference0List(lp, rp)
+      case (Literal(_), ListPat(_) | PositionalStruct(_, _)) =>
+        Right(left :: Nil)
+      case (ListPat(_), Literal(_) | PositionalStruct(_, _)) =>
+        Right(left :: Nil)
+      case (PositionalStruct(_, _), Literal(_) | ListPat(_)) =>
+        Right(left :: Nil)
+      case (WildCard | Var(_), PositionalStruct(nm, ps)) =>
+        inEnv.definedTypeFor(nm) match {
+          case None => Left(NonEmptyList.of(UnknownConstructor(nm, right, inEnv)))
+          case Some(dt) =>
+            dt.constructors.traverse {
+              case (c, params, _) if (dt.packageName, c) == nm =>
+                /*
+                 * At each position we compute the difference with _
+                 * then make:
+                 * Struct(d1, _, _), Struct(_, d2, _), ...
+                 */
+                def poke[M[_]: Applicative, A](items: List[A])(fn: A => M[List[A]]): M[List[List[A]]] =
```

Comment:
TODO: as a follow up, this can be handled recursively: we already have handling below. I'm also suspect of this method.